lightning-transaction-sync/src/electrum.rs

```diff
@@ -0,0 +1,414 @@
+					{
+						let prob_conf_height = filtered.height as u32;
+						let confirmed_tx = self.get_confirmed_tx(tx, prob_conf_height)?;
+						confirmed_txs.push(confirmed_tx);
+					}
+				}
+
+				for (i, script_history) in output_results.iter().enumerate() {
+					for possible_output_spend in script_history {
+						if possible_output_spend.height <= 0 {
+							continue;
+						}
+
+						let txid = possible_output_spend.tx_hash;
+						match self.client.transaction_get(&txid) {
+							Ok(tx) => {
+								let mut is_spend = false;
+								for txin in &tx.input {
+									let watched_outpoint = watched_outputs[i].outpoint
+										.into_bitcoin_outpoint();
+									if txin.previous_output == watched_outpoint {
+										is_spend = true;
+										break;
+									}
+								}
+
+								if !is_spend {
+									continue;
+								}
+
+								let prob_conf_height = possible_output_spend.height as u32;
+								let confirmed_tx = self.get_confirmed_tx(&tx, prob_conf_height)?;
+								confirmed_txs.push(confirmed_tx);
+							}
+							Err(e) => {
+								log_trace!(self.logger,
+									"Inconsistency: Tx {} was unconfirmed during syncing: {}",
+									txid, e);
+								return Err(InternalError::Inconsistency);
+							}
+						}
+					}
+				}
+			}
+			Err(e) => {
+				log_error!(self.logger, "Failed to script histories: {}.", e);
+				return Err(InternalError::Failed);
+			}
+		}
+
+		// Sort all confirmed transactions first by block height, then by in-block
+		// position, and finally feed them to the interface in order.
+		confirmed_txs.sort_unstable_by(|tx1, tx2| {
+			tx1.block_height.cmp(&tx2.block_height).then_with(|| tx1.pos.cmp(&tx2.pos))
+		});
+
+		Ok(confirmed_txs)
+	}
+
+	fn get_unconfirmed_transactions(
+		&self, confirmables: &Vec<&(dyn Confirm + Sync + Send)>,
+	) -> Result<Vec<Txid>, InternalError> {
+		// Query the interface for relevant txids and check whether the relevant blocks are still
+		// in the best chain, mark them unconfirmed otherwise
+		let relevant_txids = confirmables
+			.iter()
+			.flat_map(|c| c.get_relevant_txids())
+			.collect::<HashSet<(Txid, Option<BlockHash>)>>();
+
+		let mut unconfirmed_txs = Vec::new();
+
+		for (txid, block_hash_opt) in relevant_txids {
+			if let Some(block_hash) = block_hash_opt {
+				if let Ok(tx) = self.client.transaction_get(&txid) {
+					if let Some(tx_out) = tx.output.first() {
+						let script_history = self.client.script_get_history(&tx_out.script_pubkey)?;
+						if let Some(filtered) = script_history.iter()
+							.filter(|h| h.tx_hash == txid).max_by_key(|x| x.height)
+							{
+								let prob_conf_height = filtered.height;
+								let block_header = self.client.block_header(
+									prob_conf_height as usize)?;
+								if block_header.block_hash() == block_hash {
+									// Skip if the tx is still confirmed in the block in question.
+									continue;
+								}
+							}
+					}
+				}
+				unconfirmed_txs.push(txid);
```

Comment:
Won't this unconfirm the transaction if the lookup fails or the transaction has no outputs? Guessing the latter should never happen. Does the former mean Electrum never saw it? Should that result in a `InternalError::Inconsistency`? Or is there some circumstance where we would get here?